sql/core/src/test/scala/org/apache/spark/sql/CollationStringExpressionsSuite.scala

```diff
@@ -959,6 +959,37 @@ class CollationStringExpressionsSuite
     assert(collationMismatch.getErrorClass === "DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE")
   }
 
+  test("DateFormat string expression with collation") {
+    case class DateFormatTestCase[R](d: String, f: String, c: String, r: R)
+    val testCases = Seq(
+      DateFormatTestCase("2021-01-01", "yyyy-MM-dd", "UTF8_BINARY", "2021-01-01"),
+      DateFormatTestCase("2021-01-01", "yyyy-dd", "UTF8_BINARY_LCASE", "2021-01"),
+      DateFormatTestCase("2021-01-01", "yyyy-MM-dd", "UNICODE", "2021-01-01"),
+      DateFormatTestCase("2021-01-01", "yyyy", "UNICODE_CI", "2021")
+    )
+
+    testCases.foreach(t => {
```

Comment:
instead of writing all cases manually maybe think about doing this like:

```scala
for {
  collateFirstArgument <- Seq(true, false)
  collateSecondArgument <- Seq(true, false)
} {
  val firstArgument = if (collateFirstArgument) s"collate('${t.d}', '${t.c}')" else s"'${t.d}'"
  ...
}
```